CIFAR10 refactoring

ejls opened the pull request:
I've kept the format the same (the images are float32 between 0 and 255, the labels are uint8). Let me know if I missed something.

vdumoulin commented:
I would keep the dtype as `uint8` and automatically convert to `float32` and scale between 0 and 1 in `get_data`, like explained [here](https://github.com/bartvm/fuel/pull/95#issuecomment-98373974).

I would also store images in `bc01` instead of as flattened vectors for the sake of clarity: it will allow you to tag axes so that users can easily know which representation is used. You'll have to revert to using the `flatten` argument instead of `reshape`.

Once these two comments are addressed, the Scrutinizer errors are fixed and the Travis build passes, it'll be ready to merge.

vdumoulin commented:
Oh, one more thing: you can safely remove [these lines](https://github.com/ejls/fuel/blob/master/.travis.yml#L34-L44) now that `CIFAR10` is available through `fuel-download` and you're making use of it in `.travis.yml`.

vdumoulin commented:
I cleared the Travis cache, once you push a fix for the failing Scrutinizer tests the Travis tests should pass.

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +35,29 @@ class CIFAR10(IndexableDataset):
+                                                  request=request))
+        if 'features' in self.sources:
+            i = self.sources.index('features')
+            rval[i] = rval[i].astype(theano.config.floatX)
```
Use `config.floatX` here.

vdumoulin commented:
There's a Scrutinizer error remaining.

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +33,29 @@ class CIFAR10(IndexableDataset):
-                if six.PY3:
-                    batch = cPickle.load(f, encoding='latin1')
-                else:
-                    batch = cPickle.load(f)
```
In order to fix the current test failure under Python 3, you should have a look at the above code.

dwf commented:
Is there any way to get the original uint8 pixels?

vdumoulin commented:
Not ATM, but I'm working on a PR (#90) that should fix that. Basically, I'd like to introduce an attribute of datasets that's just a sane default transformer to apply. Users would be free to apply that default transformer or use the data as-is if it's what they want.

ejls commented:
@vdumoulin  thanks for the walkthrough.

I've added a parameter to select the dtype… Maybe I shouldn't have ? Currently, the pixels are NOT scaled in [0,1] since the old code wasn't doing it, should I change this as well?

vdumoulin commented:
@ejls Here's my intention for this PR:
* Converter code stores images as `uint8` between 0 and 255
* Dataset code (more precisely the `get_data` method) scales batches between 0 and 1 and casts as `fuel.config.floatX`.

I have something coming up that will remove the need for hacky `get_data` overriding and provide more flexibility, but that'll get fixed later.

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +33,31 @@ class CIFAR10(IndexableDataset):
+        kwargs.setdefault('load_in_memory', True)
         self.which_set = which_set
-        self.flatten = flatten
+        self.reshape = flatten
```
Why not use `self.flatten`?

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +33,31 @@ class CIFAR10(IndexableDataset):
         self.which_set = which_set
-        self.flatten = flatten
+        self.reshape = flatten
+        self.dtype = dtype
```
No need to make this an attribute. For the moment it's sufficient that you systematically use `fuel.config.floatX`, so you can just use that directly further down.

ejls commented:
There is already [these lines](https://github.com/bartvm/fuel/blob/master/fuel/datasets/hdf5.py#L109), so I suppose there is actually no need for flatten, I changed the test so that it uses the `H5PYDataset.__init__` parameters instead.
So it looks like this :
``` 
-    train = CIFAR10('train', load_in_memory=False)
+    train = CIFAR10('train', flatten=('features',),
+                    load_in_memory=False)
-    test = CIFAR10('test', flatten=False, load_in_memory=False)
+    test = CIFAR10('test', load_in_memory=False)
```

@vdumoulin can you clear the cache before I amend the commit? I changed the converter (again).

vdumoulin commented:
What exactly have you changed in the converter?

ejls commented:
https://github.com/ejls/fuel/blob/master/fuel/converters/cifar10.py#L39-L42 and https://github.com/ejls/fuel/blob/master/fuel/converters/cifar10.py#L56-L59

vdumoulin commented:
I just reset the cache.

By the way, it makes reviewing easier if you leave some traces of what you're doing. I get that you want to keep the commit history clean, but this is something that can be done right before merging, when I give my OK.

dwf commented:
In the case of CIFAR10 (and other colour images datasets) I'd say the sane
default is to cast to floatx and subtract the mean. Mapping to the unit
interval is useful for MNIST but kinda weird for RGB.

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +33,31 @@ class CIFAR10(IndexableDataset):
+    filename = 'cifar10.hdf5'
 
-    def __init__(self, which_set, flatten=True, **kwargs):
+    def __init__(self, which_set, flatten=True, dtype=config.floatX,
```
You should remove the `flatten` kwarg, since `H5PYDataset` already has one (which accepts a tuple of source names; that means you'll have to adapt your unit tests).

vdumoulin commented:
@dwf At the moment it can't easily be done. Like I said, there's a PR to improve this side of things, so things are bound to change anyways. For this PR I'd be happy that pixel values are mapped to the unit interval, keeping in mind that this is rather temporary.

vdumoulin commented on fuel/datasets/cifar10.py:
```diff
@@ -40,46 +33,31 @@ class CIFAR10(IndexableDataset):
+                                                  request=request))
+        if 'features' in self.sources:
+            i = self.sources.index('features')
+            rval[i] = rval[i].astype(self.dtype)
```
You should remove `self.dtype` from the constructor and instead use `config.floatX` here. Also, please divide by 255.0 before casting.

The two lines below should be removed, as `H5PYDataset` should take care of the `flatten` constructor argument by itself.

vdumoulin commented:
@ejls Thank you, looks good to merge.
